src/eventSourcing/EventSourcing.js

```diff
@@ -0,0 +1,188 @@
+
+  const endingEvent = getEndingEvent(events)
+
+  if (!endingEvent) {
+    return Date.now() - startingEvent.createdAt
+  }
+
+  return endingEvent.createdAt - startingEvent.createdAt
+}
+
+const getActiveTarget = (wf, events, inputNodeId) => {
+  // we need to find the latest phase that started and defines a command for the given
+  // input node; because the first phase in a workflow should always define default inputs
+  // for all targets, there should always be a result!
+  const phaseStartedEvents = getEventsByType(events, Event.TYPES.PHASE_STARTED).reverse()
+
+  for (const event of phaseStartedEvents) {
+    const { data: { phaseId } } = event
+    const targetValue = Workflow.getTargetValue(wf, phaseId, inputNodeId)
+
+    if (targetValue !== undefined) return targetValue
+  }
+}
+
+const getPhaseDuration = (wf, events, phaseId, occurenceIdx = 0) => {
+  const phaseStartedEvents = events.filter(({ type, data }) => (
+    type === Event.TYPES.PHASE_STARTED && data.phaseId === phaseId
+  ))
+
+  const phaseStartedEvent = phaseStartedEvents.length < (occurenceIdx + 1)
+    ? null
+    : phaseStartedEvents[occurenceIdx]
+
+  if (!phaseStartedEvent) return 0
+
+  // a phase can end in multiple ways:
+  // 1. The process is finished after it
+  // 2. One of its interrupting event listeners is triggered after it has started
+
+  const listenerIds = Workflow.getPhaseEventListeners(wf, phaseId)
+    .filter(el => el.interrupting === true)
+    .map(listener => listener.id)
+
+  const phaseEndingEvents = events.filter(({ type, data, createdAt }) => (
+    (type === Event.EVENT_RECEIVED
+      && listenerIds.includes(data.eventListenerId)
+      && createdAt >= phaseStartedEvent.createdAt)
+  ))
+
+  const phaseEndingEvent = phaseEndingEvents.length < (occurenceIdx + 1)
+    ? null
+    : phaseEndingEvents[occurenceIdx]
+
+  if (phaseEndingEvent) return phaseEndingEvent.createdAt - phaseStartedEvent.createdAt
+
+  const processEndingEvent = getEndingEvent(wf, events)
+  if (processEndingEvent) return processEndingEvent.createdAt - phaseStartedEvent.creaatedAt
+
+  return Date.now() - phaseStartedEvent.createdAt
+}
+
+const getExecutedPhaseSequence = (wf, events) => (
+  // this function does not consider that the phases might have been started
+  // on different paths of the process, that means that the n+1th phase
+  // might have been started while the nth phase is still running.
+  // But it does mean that the n+1th phase commands
+  // might have (partially overwritten) nth phase commands;
+  // it is usefull to derive things like "which target was active at which time"
+  getEventsByType(events, Event.TYPES.PHASE_STARTED)
+    .map(({ data: { phaseId } }) => Workflow.getElementById(wf, phaseId))
+)
+
+const getActivationEventsForElement = (events, el) => {
+  const { elementType, id } = el
+
+  return getEventsByType(events, ACTIVATION_EVENTS[elementType])
+    .filter(ev => ev.data[ELEMENT_TYPE_TO_ID_KEY[elementType]] === id)
+}
```

Comment:
By "Activation" I mean when the workflow has reached that element. This is used to understand what elements were already "activated" either during a running workflow, or - when user hover over an element - which elements where activated before the element was reached. This makes it possible to trace execution paths.